feat(bunny_storagezone): add a storagezone resource

jspizziri opened the pull request:
https://github.com/simplesurance/terraform-provider-bunny/issues/62

fho commented:
>  One question I have is about using a nested package/module structure (again, I'm a go noob, so I'm not sure what the language is for it). But if it were up to me I'd probably try to reduce the flatness of the internal/provider/* folder, but perhaps that's just the Go way?

In other Go projects I end up with much more packages. :-)
So far I did not see an advantage of introducing more packages.
Other provider follow the same one-package layout, see e.g. 
- https://github.com/cloudflare/terraform-provider-cloudflare/tree/master/internal/provider
- https://github.com/hashicorp/terraform-provider-kubernetes/tree/main/kubernetes

I think having all resources in the same package is fine they belong together and I don't see an advantage of splitting them to multiple packages.
Some of the helper functions could be grouped in different packages (structure.go, types_*.go) though, not sure if would really improve it.

I'm open for improving the structure. If you want to do that, please create a separate PR for it. :-)

jspizziri commented:
I'm fine with it based on your comments.

jspizziri commented:
@fho I just added tests. I think this is ready for your review. All of it of course is dependent upon the update to the `bunny-go` client. So there would need to be another commit updating the vendor folder for that package before merging at the very least.

jspizziri commented:
@fho rebased on `main`.

jspizziri commented:
@fho just cleaned up the linter errors.

jspizziri commented:
I just pushed a very small update: add the ability to import resources by id.

fho commented:
@jspizziri Sorry, I haven't had time to review it yet. 
I hope I can test + review it latest on Tuesday.
As soon as it is in master, I can create a new release.

 > I can remove the import and put it in the other PR if that makes it easier.

The smaller the PR the better. If there would be an issue with the import functionality we could still merge the rest, if it is in a separate PR.
Btw, we need a testcase for the import functionality :-)

jspizziri commented:
Import removed from this PR. I'll add it to another.

fho reviewed (commented):
@jspizziri looks great! I added some comments.

I noticed that forceRecreate does not work well with the storage zones because their deletion is postponed.
When an forceRecreate attribute is changed, "apply" will always fail because the old storage zone was not deleted yet and a new one with the same name can not be created.
I don't what a good way is to solve that.
Deletion seems to take a long time, the deleted storage zones I created still exist after ~>1h,  so using a WaitForState func would take too long.
We could always append automatically a uuid to the storage-zone name on creation.
What do you think?

Feel free to also regenerate + commit the updated docs `make docs"` + to add an entry to the changelog, otherwise I can also do it later.

jspizziri commented on internal/provider/resource_storagezone.go:
```diff
@@ -0,0 +1,326 @@
+				Sensitive:   true,
+			},
+
+			keyLastUpdated: {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```
I can't think of one off the top of my head. I added it based on the pullzone resource. LMK what you want me to do with it.

fho replied:
Lets remove it :-)

jspizziri commented:
@fho 

I've pushed a commit (which I'll squash later) that addresses your feedback with the following exceptions:

1. I couldn't find a good way to test the `ForceNew` in the `CustomizeDiff`. LMK if you have a clear path to this.
2. The storage zone recreation issue (which I'll address in long-form below)

#### Storage Zone Deletion/Recreation

I'm not in favor of appending a UUID to the end of the name because it's not what is actually specified as the name. As an example, we rely on a specific naming convention for my storage zone names, and appending something would throw that off. IMHO, this is just an eccentricity of Bunny which users should be aware of if they're using Bunny. I think if the provider is responsible for anything it would be a custom error message when failing due to an attempt to recreate a Zone if it already exists and is in the `Deleting` state (which is a property returned on the response of a Storage Zone).

I think the bigger concern with Deletes in general on a Storage Zone is that it's highly destructive. As, they're not simply deleting a zone, but also potentially GB's of data that is stored in that bucket. All of this seems to fall in the category of "know what you're doing when you're using bunny API's".

Or perhaps we should just not allow for recreations and treat all such properties as immutable?

fho commented:
> I'm not in favor of appending a UUID to the end of the name because it's not what is actually specified as the name. 

ok, I'm also not convinced about altering the name.

> I think if the provider is responsible for anything it would be a custom error message when failing due to an attempt to recreate a Zone if it already exists and is in the Deleting state (which is a property returned on the response of a Storage Zone).

Ok, also fine with me that the user has to handle it.

I wonder if it is possible to represent the deletion behavior as it is on bunny.
When a Storage-Zone is deleted, the deleted attribute is set, the storage zone (it's ID) is kept.
When ReadContext is called and storage zone does not exist we unset the ID and the storage zone vanishes.

> I think the bigger concern with Deletes in general on a Storage Zone is that they're highly destructive actions. As, they're not simply deleting a zone, but also potentially GB's of data that is stored in that zone. All of this seems to fall in the category of "know what you're doing when you're using bunny API's".
> Or perhaps we should just not allow for recreations and treat all such properties as immutable?

Do you have a specific idea how the resource would be made immutable? By returning an error from DeleteContext()? Or is there are a better way?

I googled a bit and see 2 options:

1. Without requiring any support by the provider, users can set the [lifecycle.prevent_destroy](https://www.terraform.io/language/meta-arguments/lifecycle#prevent_destroy) attribute manually on each of the storage zone, to prevent accidental deletions. (see also https://github.com/hashicorp/terraform/issues/24658). Therefore we could make this responsibility of the user.
2. The google provider prevents accidental bucket deletions, which contain objects, by having a [`force_destroy`](https://github.com/hashicorp/terraform-provider-google/blob/05f3436ceef52b4bc855f526ada80f49d26a4b24/google/resource_storage_bucket.go#L708) attribute on the resource. Only if it is set to `true` a bucket can be deleted.
We could have a similar attribute, called `allow_deletion` or so and return an error from `DeleteContext()` if it is not set.

Both options are reasonable for me. I slightly prefer 2 because of the additional safety.
What do you think?

fho commented:
@jspizziri sounds great, let's go with your suggestion :tada: 
It makes it safe and we can also document in the error message how to solve it (delete resource & define a new one with a different name) to make it clear.

jspizziri commented:
I just pushed commits for the docs generation and the changelog update.

fho commented:
@jspizziri the integration tests (`make testacc`) are failing:

```
    resource_storagezone_test.go:112: Step 1/2 error: Error running pre-apply refresh: exit status 1

        Error: 2 errors occurred:
        	* 'name' is immutable and cannot be changed from '' to 'tf-test-20220615075051824800000019'. If you must change the 'name' of our region you must first delete your resource and then redefine it. WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!
        	* 'region' is immutable and cannot be changed from '' to 'DE'. If you must change the 'region' of our region you must first delete your resource and then redefine it. WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!



          with bunny_storagezone.mytest1,
          on terraform_plugin_test.tf line 2, in resource "bunny_storagezone" "mytest1":
           2: resource "bunny_storagezone" "mytest1" {

--- FAIL: TestAccStorageZone_basic (0.19s)
=== RUN   TestAccStorageZone_full
    resource_storagezone_test.go:159: Step 1/2 error: Error running pre-apply refresh: exit status 1

        Error: 2 errors occurred:
        	* 'name' is immutable and cannot be changed from '' to 'tf-test-2022061507505201760000001a'. If you must change the 'name' of our region you must first delete your resource and then redefine it. WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!
        	* 'region' is immutable and cannot be changed from '' to 'DE'. If you must change the 'region' of our region you must first delete your resource and then redefine it. WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!



          with bunny_storagezone.mytest1,
          on terraform_plugin_test.tf line 2, in resource "bunny_storagezone" "mytest1":
           2: resource "bunny_storagezone" "mytest1" {

--- FAIL: TestAccStorageZone_full (0.30s)
```

fho commented on internal/provider/resource_storagezone.go:
```diff
@@ -0,0 +1,397 @@
+	}
+
+	updateErr := clt.StorageZone.Update(ctx, id, storageZone)
+	if updateErr != nil {
+		// if our update failed then revert our values to their original
+		// state so that we can run an apply again.
+		revertErr := revertUpdateValues(d)
+
+		if revertErr != nil {
+			return diagsErrFromErr("updating storage zone via API failed", revertErr)
+		}
```
This is not needed, the d.Get calls returns the proposed data.
The current resource state is never changed.

fho reviewed (commented):
@jspizziri looks great

I fixed all the comments that I added already locally on my own.
You can ignore them. :-)

I don't have permissions to push to you repository.
I'll create a PR for your PR with some small changes instead. If you are fine with them, we can merge it to master.

jspizziri commented:
Thanks for doing that!

jspizziri commented on internal/provider/resource_storagezone.go:
```diff
@@ -0,0 +1,397 @@
+func storageZoneFromResource(d *schema.ResourceData) (*bunny.StorageZoneUpdateOptions, error) {
+	var res bunny.StorageZoneUpdateOptions
+
+	res.ReplicationRegions = getStrSetAsSlice(d, keyReplicationRegions)
+
+	if d.HasChange(keyOriginURL) {
+		res.OriginURL = getStrPtr(d, keyOriginURL)
+	}
+
+	if d.HasChange(keyCustom404FilePath) {
+		res.Custom404FilePath = getStrPtr(d, keyCustom404FilePath)
+	}
+
+	if d.HasChange(keyRewrite404To200) {
+		res.Rewrite404To200 = getBoolPtr(d, keyRewrite404To200)
+	}
+
+	return &res, nil
```
I did it this way due to some finicky behavior where the values were getting defaulted to an empty string if empty rather than `nil` which causes a bunny error. So as long as that still happens I'm all for it.

fho replied:
I fixed those issues by using getOk instead get when retrieving the values from the resource (https://github.com/5-stones/terraform-provider-bunny/pull/1/commits/773b9beca8d5ae51bf7b3279f019e6dceedcb96c).

jspizziri commented:
Based on your earlier comment it sounds like you've already fixed the issues you mentioned. But if you need me to fix anything just let me know and I'll hop on it.

jspizziri commented:
I've reached out to Bunny Support about "OriginUrl". I'm hoping we can just remove it.

fho commented:
@jspizziri feel free to revert my ` storagezone: remove date_modified key ` commit if you prefer.
I did not realize that this is actually a field returned by the bunny api. I thought it's a field that the provider computates. 

I don't know if there is a usecase for the field in the provider :shrug:.

jspizziri commented on internal/provider/resource_storagezone.go:
```diff
@@ -0,0 +1,399 @@
+				return nil
+			}),
+			customdiff.ValidateChange(keyCustom404FilePath, func(_ context.Context, old interface{}, new interface{}, meta interface{}) error {
+				return validateUnremovableProperty(keyCustom404FilePath, old, new)
```
That makes good sense.

jspizziri replied:
@fho I made this change and squashed it into the previous commit that added the default and validation for the properties originally. So now it's just one commit adding defaults for both properties.

https://github.com/simplesurance/terraform-provider-bunny/pull/64/commits/69eb1c6740f09898bbcc1db62567c174c96a47d6

jspizziri commented on internal/provider/resource_storagezone.go:
```diff
@@ -0,0 +1,399 @@
+		return diagsErrFromErr("updating storage zone via API failed", updateErr)
+	}
+
+	updatedStorageZone, err := clt.StorageZone.Get(ctx, id)
+	if err != nil {
+		return diagsErrFromErr("fetching updated storage zone via API failed", err)
+	}
+
+	if err := storageZoneToResource(updatedStorageZone, d); err != nil {
+		return diagsErrFromErr("converting api type to resource data after successful update failed", err)
+	}
```
So I did that because looking at the pull zone as a template that update returns a response containing the pz, which is then fed back into the state update function. Because SZ update returns a 204 I assumed I would need to read the SZ after update to write it. But in retrospect I think you're right and that TF already knows the new state.

jspizziri replied:
@fho I pushed a commit that removes this block.

jspizziri commented:
> @jspizziri feel free to revert my ` storagezone: remove date_modified key ` commit if you prefer.
> 
> I did not realize that this is actually a field returned by the bunny api. I thought it's a field that the provider computates. 
> 
> 
> 
> I don't know if there is a usecase for the field in the provider :shrug:.

Yeah I don't really think it's useful. Especially since it will be constantly out of date unless you're refreshing your state. I'm fine leaving it out.

jspizziri commented:
@fho alright, I think this is ready for another round of review:

1. Your commit has been merged.
2. I removed the unnecessary read/write to tf state on update
3. I added the default for the 404 URL that we discussed.

fho commented:
@jspizziri  

> I've reached out to Bunny Support about "OriginUrl". I'm hoping we can just remove it.

Do you got a response from the bunny support?
I created https://github.com/simplesurance/terraform-provider-bunny/issues/77 to track the issue.

The OriginURL issue is only a minor thing and everything else looks fine so I finally merged the PR :partying_face:.
Thanks a lot for your contribution and the great team work. :-)

I'll create a new release presumably end of the week.

jspizziri commented:
I haven't seen a response yet but I'll let you know if I do.

Awesome, thanks a ton for all your help and work on getting this across the finish line!

Looking forward to using it 😎
